File: suites/osprey-compute-kernels/reference/cases/isqrt/isqrt.ml
(* Sum of integer square roots over 1..N — Newton's method, integer division heavy. *)
let rec isqrt_iter n x =
  let y = (x + n / x) / 2 in
  if y < x then isqrt_iter n y else x

let isqrt n = if n < 2 then n else isqrt_iter n n

let () =
  let acc = ref 0 in
  for i = 1 to 1000000 do
    acc := !acc + isqrt i
  done;
  Printf.printf "%d\n" !acc

File: suites/osprey-compute-kernels/reference/cases/isqrt/isqrt.rs
// Sum of integer square roots over 1..N — Newton's method, integer division heavy.
fn isqrt(n: i64) -> i64 {
    if n < 2 {
        return n;
    }
    let mut x = n;
    loop {
        let y = (x + n / x) / 2;
        if y < x {
            x = y;
        } else {
            return x;
        }
    }
}

fn main() {
    let mut acc: i64 = 0;
    let mut i: i64 = 1;
    while i < 1000001 {
        acc += isqrt(i);
        i += 1;
    }
    println!("{}", acc);
}
